### PR TITLE
sketch of a more invasive CompletableFuture patch

### PR DESCRIPTION
**Pull request by mosesn**
This more invasive approach (compared to #121) lets us delete a lot of code.  I think it's a matter of taste whether it's better or not, but I figured this would illustrate what I meant more clearly than chatting over PRs :).

I do not think we should merge this as-is.  In particular, I put a lot of logic in an anonymous lambda in the constructor, and we can likely find a better home for it.  I would also completely understand you saying that you don't like this approach and simply closing the PR.

### CODE REVIEW

**Comment by vorburger** [comment]
@mosesn this looks interesting! I'm in principle in favour of something like this, because I like the code simplification that this seems to bring. I would need to review it in full details though. And we're both so excited that both of our work overlaps a little bit - which is totally OK! Could I propose the following next steps in this order:

1. Do you want to review my #123 + #124 and I merge those, for now? I do understand that this PR will then remove that again - and that's totally fine with me! I just prefer getting things merged in order, and then change them again (and have a baseline to revert back to, in the unlikely event that we would have to back this out again later).

2. I finish a PR that I have WIP to add Checkstyle, in order to settle [tabs vs spaces](https://www.youtube.com/watch?v=SsoOG6ZeyUI) 🤣 

3. [I go to sleep]

4. You rebase this PR

5. I review this PR in details, may ask some questions and feedback, and then merge it

Would this sequence be OK for you?

**Comment by mosesn** [comment]
🤦 Sorry about the tabs, I haven't set up my emacs config properly for Java on my personal laptop and I try to avoid opening my work laptop on the weekend . . . I'll get right to that.  That approach sounds good to me!  One nit: how about we use spotless instead of checkstyle?  Spotless has an autofixer that doesn't mean I have to use IntelliJ haha

**Comment by vorburger** [comment]
@mosesn _voilà_ steps 1. & 2. are done, I'll go to 😴  now and this is all yours for 4!

**Comment by mosesn** [comment]
Looks like there are still a couple of bugs in it.  Interestingly, I get a couple failures locally but they're different ones:

```
[ERROR] Failures:
[ERROR]   ManagedProcessTest.onProcessFailedInvokedOnCustomListenerWithExitValueChecker:67 expected:<2> but was:<1>
[ERROR] Errors:
[ERROR]   ManagedProcessTest.mustTerminateExec:210 » ManagedProcess Program [sleep, 30s] (in working directory /Users/moses/projects/ch.vorburger.exec/.) failed with Exception: , last 100 lines of console:
usage: sleep seconds
```
I think it's just discrepancies in the `sleep` and `ls` implementations though.  I'll have to do some poking to figure out what's going on with the other issues.  I don't think I'll get to it tonight, I'll try to figure it out later this week.

**Comment by vorburger** [comment]
> I think it's just discrepancies in the `sleep` and `ls` implementations though.  

Are you on Mac locally? The tests on the main branch may well be broken on Mac. (The CI only covers Linux.)

If so, raise a small separate PR to fix that, first?

**Comment by mosesn** [comment]
> > I think it's just discrepancies in the `sleep` and `ls` implementations though.
> 
> Are you on Mac locally? The tests on the main branch may well be broken on Mac. (The CI only covers Linux.)
> 
> If so, raise a small separate PR to fix that, first?

#135

**Comment by vorburger** [comment]
@mosesn Hi! It's been a while... 😃 Do you want to rebase this, and we can get this in?

**Comment by mosesn** [comment]
Hey, I think I'll get some time in the next few weeks to take a stab at rebasing this!  Sorry for the delay

**Comment by vorburger** [comment]
@mosesn did you still want to rebase and clean this up for re-review and possible inclusion? Or forget about & close this one?

**Comment by vorburger** [comment]
@mosesn ping? 👋🏽

**Comment by mosesn** [comment]
Sorry for the delay! Finally found some time to work on OSS stuff while flying 🙃

File: src/main/java/ch/vorburger/exec/ManagedProcess.java [hunk 1]
```diff
@@ -125,8 +129,25 @@ public class ManagedProcess implements ManagedProcessState {
         this.destroyOnShutdown = destroyOnShutdown;
         this.consoleBufferMaxLines = consoleBufferMaxLines;
         this.outputStreamLogDispatcher = outputStreamLogDispatcher;
-        this.resultHandler = new CompositeExecuteResultHandler(this,
-                Arrays.asList(new LoggingExecuteResultHandler(this), new ProcessResultHandler(listener)));
+        this.asyncResult = new CompletableFuture<>();
+        this.asyncResult.<Void>handle((result, e) ->
```
[the remark this answers is not in the document]
Comment (mosesn) [reply src/main/java/ch/vorburger/exec/ManagedProcess.java]: @sonatype-lift ignore

**Comment by mosesn** [comment]
@vorburger ready for your review when you get a chance!

**Comment by vorburger** [comment]
LGTM! Thank You so much for having gotten back to this (it's been a while). Merging ASAP!

**Review by vorburger — APPROVED** [review]

**Pull request merged**